src/structs/probability_tables.rs

```diff
@@ -240,40 +241,45 @@ impl ProbabilityTables {
         // here DC in raster_cols should be 0
         let pixels_sans_dc = run_idct(raster_cols);
 
-        // helper functions to avoid code duplication that calculate the left and above prediction values
-
-        let calc_pred = |init_pred: i16x8, a1: i16x8, a2: i16x8| {
-            if enabled_features.use_16bit_adv_predict {
+        // helper functions to avoid code duplication that calculate prediction values
+        #[inline]
+        fn calc_pred(a1: i16x8, a2: i16x8, is_16_bit: bool) -> i16x8 {
+            if is_16_bit {
                 let pixel_delta = a1 - a2;
                 let half_delta = (pixel_delta - (pixel_delta >> 15)) >> 1; /* divide pixel_delta by 2 rounding towards 0 */
 
-                init_pred - a1 - 128 * X_IDCT_SCALE as i16 - half_delta
+                a1 + half_delta
             } else {
                 let a1 = i32x8::from_i16x8(a1);
                 let a2 = i32x8::from_i16x8(a2);
                 let pixel_delta = a1 - a2;
                 let half_delta = (pixel_delta - (pixel_delta >> 31)) >> 1; /* divide pixel_delta by 2 rounding towards 0 */
-                let result = i32x8::from_i16x8(init_pred) - a1 - 128 * X_IDCT_SCALE - half_delta;
+                let result = a1 + half_delta;
 
                 i16x8::from_i32x8_truncate(result)
             }
-        };
+        }
 
-        let calc_left = || {
-            let left_pred = neighbor_data.neighbor_context_left.get_vertical_pix();
-            let a1 = pixels_sans_dc.from_stride(0, 8);
-            let a2 = pixels_sans_dc.from_stride(1, 8);
+        // transpose so we can get the vertical rows as single vectors
+        let transposed = pixels_sans_dc.transpose();
```

Comment:
Is it faster than previous variant? Transposition is quite heavy operation.

Reply:
You can use partial transpose to speedup the things a bit: only two final rows are needed.

Reply:
I need to check again, but it looked like the optimizer already discards some of the transpose operations since the outputs aren't used.

Reply:
I was right (not very happy about though)

Current `main` on my box gives
```
 Performance counter stats for 'taskset -c 10 nice -n -20 target/release/lepton_jpeg_util images/img_52MP_7k.lep images/img_52MP_7k.jpg':

       438 105 179      cache-references                                                        (41,57%)
        37 118 112      cache-misses                     #    8,47% of all cache refs           (41,36%)
     8 291 275 273      cycles                                                                  (41,65%)
       447 758 283      ic_fetch_stall.ic_stall_back_pressure                                        (41,95%)
       495 647 961      stalled-cycles-frontend          #    5,98% frontend cycles idle        (42,23%)
    19 066 223 320      instructions                     #    2,30  insn per cycle            
                                                  #    0,03  stalled cycles per insn     (42,29%)
     1 782 521 906      branch-instructions                                                     (42,40%)
        68 665 185      branch-misses                    #    3,85% of all branches             (42,28%)
     3 802 966 088      ic_fetch_stall.ic_stall_any                                             (42,09%)
        23 575 889      ic_fetch_stall.ic_stall_dq_empty                                        (41,83%)
        29 695 256      l2_cache_misses_from_ic_miss                                            (41,90%)
     1 017 503 279      l2_latency.l2_cycles_waiting_on_fills                                        (41,65%)
            99 462      faults                                                                
                 1      migrations                                                            

       1,848447492 seconds time elapsed

       1,650808000 seconds user
       0,194741000 seconds sys
```
And this PR is slower
```
 Performance counter stats for 'taskset -c 10 nice -n -20 target/release/lepton_jpeg_util images/img_52MP_7k.lep images/img_52MP_7k.jpg':

       433 451 044      cache-references                                                        (41,70%)
        34 034 653      cache-misses                     #    7,85% of all cache refs           (41,50%)
     8 384 320 430      cycles                                                                  (41,59%)
       416 005 396      ic_fetch_stall.ic_stall_back_pressure                                        (41,71%)
       498 876 200      stalled-cycles-frontend          #    5,95% frontend cycles idle        (41,82%)
    19 313 060 705      instructions                     #    2,30  insn per cycle            
                                                  #    0,03  stalled cycles per insn     (42,09%)
     1 814 232 887      branch-instructions                                                     (42,43%)
        66 979 824      branch-misses                    #    3,69% of all branches             (42,43%)
     3 842 459 194      ic_fetch_stall.ic_stall_any                                             (42,12%)
        20 625 976      ic_fetch_stall.ic_stall_dq_empty                                        (41,91%)
        26 322 984      l2_cache_misses_from_ic_miss                                            (41,98%)
     1 054 282 849      l2_latency.l2_cycles_waiting_on_fills                                        (41,91%)
            99 479      faults                                                                
                 1      migrations                                                            

       1,865674078 seconds time elapsed

       1,681903000 seconds user
       0,180774000 seconds sys
```

Reply:
So I suggest to remove transposition and get back to `from_stride` operations.